### PR TITLE
MG-68 - Switch to GitHub Actions

### LINKED ISSUE #35: Improve CI speed by using self-hosted agnet
**ENHANCEMENT**
1. Describe the enhancement you are requesting. 
This is a tooling feature with the goal of improvement of our CI cycle.
The idea is to set up a [Semaphore CI Agent](https://github.com/semaphoreci/agent) like described [here](https://docs.semaphoreci.com/ci-cd-environment/install-self-hosted-agent/).

2. Indicate the importance of this enhancement to you (must-have, should-have, nice-to-have).
This is a must-have since our CI is getting painfully slow with limited support for parallelism.

### LINKED ISSUE #68: Switch to GitHub Actions
**ENHANCEMENT**
1. Describe the enhancement you are requesting.
Use GitHub actions for CI.

2. Indicate the importance of this enhancement to you (must-have, should-have, nice-to-have).
This is a should-have.

### PR DESCRIPTION
**Pull request by Musilah**
### What does this do?
Improve CI speed by setting up a Github Actions Runner
### Which issue(s) does this PR fix/relate to?
Resolves #68, related to #35.

### List any changes that modify/break current functionality
None
### Have you included tests for your changes?
None
### Did you document any new/modified functionality?
No

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,1555 @@
...
+            ${{ runner.os }}-codecov-
+
+      - name: Setup Golangci-lint
+        run: |
+          curl -sSfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh | sh -s v1.53.3
+          
+      - name: Cache Golangci-lint
+        uses: actions/cache@v2
+        with:
+          path: ./bin/golangci-lint
```
Comment (WashingtonKK) [inline .github/workflows/ci.yml:45]: ``
```suggestion
          path: ./bin
```

File: auth.pb.go [hunk 2]
```diff
@@ -4,7 +4,7 @@
 // Code generated by protoc-gen-go. DO NOT EDIT.
 // versions:
 // 	protoc-gen-go v1.31.0
-// 	protoc        v4.24.4
+// 	protoc        v4.24.3
```
Comment (rodneyosodo) [inline auth.pb.go:7]: Revert

File: .github/workflows/golangci-lint.yml [hunk 3]
```diff
@@ -0,0 +1,29 @@
+# Copyright (c) Abstract Machines
+# SPDX-License-Identifier: Apache-2.0
+
```
Comment (rodneyosodo) [inline .github/workflows/golangci-lint.yml:3]: This doesn't need to be in a separate file. You can just combine them

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -0,0 +1,1295 @@
...
+
+on:
+  push:
+    branches:
+      - main
+      - MG-35-self-hostrunner
+  pull_request:
+    branches:
+      - main
+      - MG-35-self-hostrunner
```
Comment (rodneyosodo) [inline .github/workflows/ci.yml:14]: no need for this at the moment. It should be on push to main or on pull request to main

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -0,0 +1,1295 @@
...
+
+jobs:
+  setup:
+    runs-on: ubuntu-20.04
+
+    steps:
+      - name: Checkout Code
+        uses: actions/checkout@v2
+
+      - name: Setup Codecov
```
Comment (rodneyosodo) [inline .github/workflows/ci.yml:24]: Why not use official action by `codecov` https://github.com/codecov/codecov-action ?

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -0,0 +1,1295 @@
...
+      - main
+      - MG-35-self-hostrunner
+  pull_request:
+    branches:
+      - main
+      - MG-35-self-hostrunner
+
+jobs:
+  setup:
+    runs-on: ubuntu-20.04
```
Comment (rodneyosodo) [inline .github/workflows/ci.yml:18]: This ubuntu version is pretty old use `ubuntu-latest`

**Review by rodneyosodo — CHANGES_REQUESTED** [review]

**Review by dborovcanin — CHANGES_REQUESTED** [review]
Since all the services are very similar, can we use a matrix with the service list instead of having repeated sections per service?

**Review by dborovcanin — CHANGES_REQUESTED** [review]
We also need to trigger actions if GH Actions files change (such as proto update or other changes in the Actions file).

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -0,0 +1,494 @@
...
+              exit 1
+            fi
+          done
+
+          for p in $(ls pkg/messaging/*.pb.go); do
+            if ! cmp -s $p $p.tmp; then
+              echo "Proto file and generated Go file $p are out of sync!"
+              exit 1
+            fi
+          done
```
Comment (arvindh123) [inline .github/workflows/ci.yml:116]: This looks good, May be this command `find . -name "*.pb.go"` will help you get all `*.pb.go` in repo in single step

File: .github/workflows/ci.yml [hunk 8]
```diff
@@ -0,0 +1,651 @@
...
+          go-version: 1.21
+
+      - name: Login to Docker Hub
+        run: echo ${{ secrets.DOCKER_TOKEN }} | docker login --username ${{ secrets.DOCKER_USERNAME }} --password-stdin
+
+      - name: Build and push Provision
+        run: |
+          docker push magistrala/http:latest
+        env:
+          DOCKER_HUB_SECRET: ${{ secrets.DOCKER_HUB }}
```
[the remark this answers is not in the document]
Comment (dborovcanin) [reply .github/workflows/ci.yml:278]: Also, can the whole Docker push be a part added to the matrix, so we don't have to repeat it for all the services?

**Review by dborovcanin — CHANGES_REQUESTED** [review]

File: .github/workflows/build.yml [hunk 9]
```diff
@@ -0,0 +1,115 @@
+# This workflow is licensed under the Apache License, Version 2.0
+# Copyright (c) Abstract Machines
+
+name: Build and Publish Docker Image
```
Comment (rodneyosodo) [inline .github/workflows/build.yml:4]: What is the difference between this and `cd` ?

**Review by rodneyosodo — CHANGES_REQUESTED** [review]

File: .github/workflows/build.yml (hunk 9 above)
Comment (Musilah) [reply .github/workflows/build.yml:4]: ah im supposed to delete it. apologies

**Pull request merged**